nutanix/resource_nutanix_project.go

```diff
@@ -569,3 +1284,504 @@ func expandMetadata(d *schema.ResourceData, kind string) *v3.Metadata {
+		val := v1.(map[string]interface{})
+		res := &v3.Reference{}
+		if v1, ok1 := val["uuid"]; ok1 && len(v1.(string)) > 0 {
+			res.UUID = utils.StringPtr(v1.(string))
+		}
+		if v1, ok1 := val["name"]; ok1 && len(v1.(string)) > 0 {
+			res.Name = utils.StringPtr(v1.(string))
+		}
+		if v1, ok1 := val["kind"]; ok1 && len(v1.(string)) > 0 {
+			res.Kind = utils.StringPtr(key)
+		}
+		return res
+	}
+	return nil
+}
+
+func UpdateACPNoChange(resp *v3.ProjectInternalIntentResponse) []*v3.AccessControlPolicyList {
+	acpList := resp.Spec.AccessControlPolicyList
+	acpRes := make([]*v3.AccessControlPolicyList, len(acpList))
+	for k, v := range acpList {
+		acps := &v3.AccessControlPolicyList{}
+
+		acps.ACP = v.ACP
+		acps.Metadata = v.Metadata
+		acps.Metadata = v.Metadata
+		acps.Operation = utils.StringPtr("UPDATE")
+
+		acpRes[k] = acps
+	}
+	return acpRes
+}
+
+func expandCreateAcp(pr []interface{}, d *schema.ResourceData, projectUUID string, clusterUUID string, meta interface{}) []*v3.AccessControlPolicyList {
+	if len(pr) > 0 {
+		acpList := make([]*v3.AccessControlPolicyList, len(pr))
+
+		for k, val := range pr {
+			acps := &v3.AccessControlPolicyList{}
+			acpSpec := &v3.AccessControlPolicySpec{}
+			acpRes := &v3.AccessControlPolicyResources{}
+
+			v := val.(map[string]interface{})
+
+			if v1, ok1 := v["name"]; ok1 {
+				acpSpec.Name = utils.StringPtr(v1.(string))
+			}
+			if v1, ok1 := v["description"]; ok1 {
+				acpSpec.Description = utils.StringPtr(v1.(string))
+			}
+			if v1, ok := v["user_reference_list"]; ok {
+				acpRes.UserReferenceList = validateArrayRef(v1.(*schema.Set), utils.StringPtr("user"))
+			}
+
+			if v1, ok := v["user_group_reference_list"]; ok {
+				acpRes.UserGroupReferenceList = validateArrayRef(v1.(*schema.Set), utils.StringPtr("user_group"))
+			}
+
+			if v, ok := v["role_reference"]; ok {
+				acpRes.RoleReference = validateRefList(v.([]interface{}), nil)
+
+				// role uuid
+				roleID := acpRes.RoleReference.UUID
+
+				// check for project collaboration. default is set to true
+				pcCollab := true
+				if pc, ok1 := d.GetOkExists("enable_collab"); ok1 {
+					pcCollab = pc.(bool)
+				}
+				// get permissions based on role
+
+				conList := getRolesPermission(*roleID, meta, projectUUID, clusterUUID, pcCollab)
+
+				// generate the filter list based on role
+				filterList := &v3.FilterList{}
+				filterList.ContextList = conList
+
+				if filterList.ContextList != nil {
+					acpRes.FilterList = filterList
+				}
+			}
+
+			metadata := &v3.Metadata{}
+			metadata.Kind = utils.StringPtr("access_control_policy")
+
+			acps.Operation = utils.StringPtr("ADD")
+			acps.Metadata = metadata
+			acpSpec.Resources = acpRes
+			acps.ACP = acpSpec
+			acpList[k] = acps
+		}
+
+		return acpList
+	}
+	return nil
+}
+
+func UpdateExpandAcpRM(pr []interface{}, res *v3.ProjectInternalIntentResponse, d *schema.ResourceData, meta interface{}, projectUUID string, clusterUUID string) []*v3.AccessControlPolicyList {
```

Comment:
What is this function doing ?
Can you add some code comments on these functions?

Reply:
This is called when we have to update the acps (adding, deleting) in UpdateContext when projects_internal flag is set.
Will add comments.